core/core-db-repositories/src/main/java/com/sflpro/notifier/db/repositories/repositories/notification/push/impl/PushNotificationRecipientRepositoryImpl.java

```diff
@@ -0,0 +1,120 @@
+        LOGGER.debug("Executing push notification recipients search request for search parameters, parameters - {}, startFrom - {} ,maxCount - {}", parameters, startFrom, maxCount);
+        // Build query
+        final TypedQuery<PushNotificationRecipient> typedQuery = buildFindRecipientsForSearchParametersTypedQuery(parameters, false, PushNotificationRecipient.class);
+        // Set pagination data
+        typedQuery.setFirstResult((int) startFrom);
+        typedQuery.setMaxResults(maxCount);
+        // Execute query
+        final List<PushNotificationRecipient> recipients = typedQuery.getResultList();
+        LOGGER.debug("Successfully executed recipients lookup request for search parameters, parameters - {}, startFrom - {} ,maxCount - {}", parameters, startFrom, maxCount);
+        return recipients;
+    }
+
+    /* Utility methods */
+    private void assertSearchParametersNotNull(final PushNotificationRecipientSearchFilter parameters) {
+        Assert.notNull(parameters, "Search parameters should not be null");
+    }
+
+    private <T> TypedQuery<T> buildFindRecipientsForSearchParametersTypedQuery(final PushNotificationRecipientSearchFilter parameters, final boolean countQuery, final Class<T> queryResultType) {
+        final String queryString = buildFindRecipientsForSearchParametersQueryString(parameters, countQuery);
+        // Create typed query
+        final TypedQuery<T> typedQuery = entityManager.createQuery(queryString, queryResultType);
+        typedQuery.setParameter(PARAMETER_NAME_SUBSCRIPTION_ID, parameters.getSubscriptionId());
+        typedQuery.setParameter(PARAMETER_NAME_PROVIDER_TYPE, parameters.getProviderType());
+        typedQuery.setParameter(PARAMETER_NAME_DEVICE_OPERATING_SYSTEM, parameters.getDeviceOperatingSystemType());
+        typedQuery.setParameter(PARAMETER_NAME_DESTINATION_ROUTE_TOKEN, parameters.getDestinationRouteToken());
+        typedQuery.setParameter(PARAMETER_NAME_STATUS, parameters.getStatus());
+        typedQuery.setParameter(PARAMETER_NAME_APPLICATION_TYPE, parameters.getApplicationType());
+        return typedQuery;
+    }
+
+    private String buildFindRecipientsForSearchParametersQueryString(final PushNotificationRecipientSearchFilter parameters, final boolean countQuery) {
+        // Build query string
+        final StringBuilder queryBuilder = new StringBuilder();
+        if (countQuery) {
+            queryBuilder.append(" select count(pnr) ");
+        } else {
+            queryBuilder.append(" select pnr ");
+        }
+        queryBuilder.append(" from PushNotificationRecipient pnr ");
+        queryBuilder.append(" where ");
+        queryBuilder.append("( :" + PARAMETER_NAME_DESTINATION_ROUTE_TOKEN + " is null or pnr.destinationRouteToken=:" + PARAMETER_NAME_DESTINATION_ROUTE_TOKEN + ") ");
+        queryBuilder.append(" and ( :" + PARAMETER_NAME_STATUS + " is null or pnr.status=:" + PARAMETER_NAME_STATUS + ") ");
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal " and ( :" 5 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)